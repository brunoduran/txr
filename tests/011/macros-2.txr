@(do
   (set *gensym-counter* 0)

   (macro-time
     (defmacro whilst ((condition : result) . body)
       (let ((cblk (gensym "cnt-blk-"))
             (bblk (gensym "brk-blk-")))
         ^(macrolet ((break (value) ^(return-from ,',bblk ,value)))
            (symacrolet ((break (return-from ,bblk))
                         (continue (return-from ,cblk)))
              (block ,bblk
                (for () (,condition ,result) ()
                  (block ,cblk ,*body))))))))

   (let ((i 0))
     (whilst ((< i 100))
       (if (< (inc i) 20)
         continue)
       (if (> i 30)
         break)
       (prinl i)))

   (prinl
     (ignwarn
       (expand
         '(whilst ((< i 100))
            (if (< (inc i) 20)
              continue)
            (if (> i 30)
              break)
            (prinl i)))))

   (let ((i 0))
     (whilst ((< i 5))
       (inc i)
       (labels ((whilst () 'whilst))
         (let ((break 'break))
           (prinl (list (whilst) break)))))))
