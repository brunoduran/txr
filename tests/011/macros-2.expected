20
21
22
23
24
25
26
27
28
29
30
(block #:brk-blk-0002
  (block () (let () (sys:for-op ()
                      ((< i 100) ())
                      () (block #:cnt-blk-0001
                           (if (< (sys:setq i (succ i))
                                  20)
                             (return-from
                               #:cnt-blk-0001))
                           (if (> i 30)
                             (return-from
                               #:brk-blk-0002))
                           (prinl i))))))
(whilst break)
(whilst break)
(whilst break)
(whilst break)
(whilst break)
